### PR TITLE
detect and visualize self calls

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Review by pontuslaestadius — DISMISSED** [review]
[https://i.stack.imgur.com/it6wg.gif](url)

Don't self calls advance the lifeline?

File: Application/src/visuals/Message.java [hunk 1]
```diff
@@ -8,7 +8,8 @@
 /**
  * Class for creating the messages to pass between "classes".
  * @author Sebastian Fransson
- * @version 1.0
+ * -collaborator Rashad Kamsheh
+ * @version 1.1
```
Comment (IsabelleTornqvist) [inline Application/src/visuals/Message.java]: Version 2.0?

Comment (ghost) [reply Application/src/visuals/Message.java]: @IsabelleTornqvist 
1.0++ is 1.1
1.9++ is 2.0

File: Application/src/visuals/Message.java [hunk 2]
```diff
@@ -85,7 +88,27 @@ else if(keepAnimating && node1.getX() > node2.getX()){ // Sending a return messa
...
            }
-    }
+            //Checks if we are supposed to keep animating, set animationBounds according to how diagramClasses are scaled.
+           else if (keepAnimating && node1.getX() == node2.getX()) { //Sending a self referencing message
+               switch (stage) {
+                   case 1 :
+                       switchDirection = true; // Switch state of the dragon's flying direction
+                       // Move 60 pixels to the right
+                       if((animationBounds += (class_size/6)) > 40) {
+                           offset += 28; // Lower the dragon vertically
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Where did you get this number from

**Review by timjon — CHANGES_REQUESTED** [review]
Should the dragon really be teleporting down?

**Review by pontuslaestadius — DISMISSED** [review]
Values don't scale. Meaning scaling the application makes it look weird. Also the current values probably only look good for 1 set resolution?

**Review by Sebastian-Fransson — COMMENTED** [review]
ATM as Tim pointed out the dragon is teleporting down in the self call. Is there a way you could make it so that the dragon decends instead of popping in and out?

File: Application/src/visuals/Message.java [hunk 3]
```diff
@@ -75,22 +83,50 @@ public boolean isKeepAnimating() {
...
+                    }
+                    break;
+
+                case 2:
+                    // Multiple increments to lower the message vertically
+                    for (int i = 0; i < 3; i++) {
+
+                        offset += 4;
+                    }
+                    stage = 3;
```
Comment (kosara-g) [inline Application/src/visuals/Message.java]: Is it intended to have no break here?

File: Application/src/visuals/Message.java [hunk 4]
```diff
@@ -75,22 +83,50 @@ public boolean isKeepAnimating() {
...
+
+                case 3:
+                    // move back to starting point
+                    if ((animationBounds -= (class_size / 6)) < 1) {
+                        keepAnimating = false;
+                    }
+                    break;
+            }
+        }
```
Comment (kosara-g) [inline Application/src/visuals/Message.java]: no default case? :(

File: Application/src/visuals/Message.java [hunk 5]
```diff
@@ -7,19 +7,23 @@
...
-    private Coordinates coordinates , node1, node2; // The coordinates of the nodes that the message is supposed to pass between.
+    private Coordinates coordinates, node1, node2; // The coordinates of the nodes that the message is supposed to pass between.
     private int fromNode, toNode;
     private int offset; // offset for the message "ordering".
     private int class_size; // identifier for the current class size so that messages can scale.
     private int animationBounds = 0; //Starts and stops the animation at the correct bounds.
     private boolean keepAnimating = true; // State of the animation. Beginning or ending.
     private boolean switchImage; // Keeps track of which image to show.
     private double messageScale = 1.5;
+    private int stage = 1; // Used for the hardcoded trajectory of self referencing message.
```
Comment (pontuslaestadius) [inline Application/src/visuals/Message.java]: Why is it called stage? Is it the same kind of stage we are currently using in javafx?

Comment (ghost) [reply Application/src/visuals/Message.java]: no, it should be renamed

File: Application/src/visuals/Message.java [hunk 6]
```diff
@@ -112,39 +147,57 @@ public void render(GraphicsContext gc){
...
+            }
+
+            // Handling the switch of the dragon's flying direction when visualising self calls
+            // Checks if up image is supposed to be shown and if the direction of flying dragon should be switched.
+            else if (switchDirection && switchImage) {
+                //sets the dimensions of the dragon according to the current class size.
+                gc.drawImage(dragonMessage, x1 + animationBounds,
+                        y1 + (this.class_size), class_size / messageScale, class_size / messageScale); //State Wings Down.
+                //TODO: uncomment the following line after integration with the trail feature
+                //this.trails.add(new Trail((x1)+ animationBounds, y1 +(this.class_size),class_size/trailScale, class_size/trailScale));
```
Comment (timjon) [inline Application/src/visuals/Message.java]: This shouldnt be here right

File: Application/src/visuals/Message.java [hunk 7]
```diff
@@ -1,25 +1,27 @@
...
  * @author Sebastian Fransson
- * @version 1.0
+ * -collaborator Rashad Kamsheh
+ * @version 1.1
  */
-public class Message implements Renderable{
+public class Message implements Renderable {
     private String name;
-    private Coordinates coordinates , node1, node2; // The coordinates of the nodes that the message is supposed to pass between.
+    private Coordinates coordinates, node1, node2; // The coordinates of the nodes that the message is supposed to pass between.
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 8]
```diff
@@ -34,7 +36,7 @@
     /**
      * Constructor
      */
-    public Message(Coordinates node1, Coordinates node2, String name, int fromNode, int toNode, int offset, int size){
+    public Message(Coordinates node1, Coordinates node2, String name, int fromNode, int toNode, int offset, int size) {
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 9]
```diff
@@ -46,13 +48,17 @@ public Message(Coordinates node1, Coordinates node2, String name, int fromNode,
 
     /**
      * Retrieves the coordinates from the Message object.
+     *
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 10]
```diff
@@ -46,13 +48,17 @@ public Message(Coordinates node1, Coordinates node2, String name, int fromNode,
...
      * @return coordinates
      */
     @Override
-    public Coordinates getCoordinates() { return coordinates; }
+    public Coordinates getCoordinates() {
+        return coordinates;
+    }
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 11]
```diff
@@ -46,13 +48,17 @@ public Message(Coordinates node1, Coordinates node2, String name, int fromNode,
...
 
     /**
      * Formatting method.
+     *
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 12]
```diff
@@ -75,22 +81,51 @@ public boolean isKeepAnimating() {
...
-                   keepAnimating = false;
-           }
-            //Checks if we are supposed to keep animating, set animationBounds according to how diagramClasses are scaled.
-           else if(keepAnimating && node1.getX() > node2.getX()){ // Sending a return message.
-               if((animationBounds -= (class_size/6 +curve)) < (this.node2.getX())  - this.node1.getX())
-                   keepAnimating = false;
-           }
+        //Checks if we are supposed to keep animating, set animationBounds according to how diagramClasses are scaled.
+        if (keepAnimating && node2.getX() > node1.getX()) { // Sending a message.
+            if ((animationBounds += (class_size / 6)) > this.node2.getX() - this.node1.getX())
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 13]
```diff
@@ -75,22 +81,51 @@ public boolean isKeepAnimating() {
...
+                keepAnimating = false;
+        }
+        //Checks if we are supposed to keep animating, set animationBounds according to how diagramClasses are scaled.
+        else if (keepAnimating && node1.getX() > node2.getX()) { // Sending a return message.
+            if ((animationBounds -= (class_size / 6)) < (this.node2.getX()) - this.node1.getX())
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java (hunk 4 above)
Comment (timjon) [reply Application/src/visuals/Message.java]: +1

File: Application/src/visuals/Message.java [hunk 14]
```diff
@@ -75,22 +81,51 @@ public boolean isKeepAnimating() {
...
+                    break;
+            }
+        }
     }
 
     /**
      * Gets the new coordinates from resizing the application and redraws messages with these coordinates.
      */
-    public void changeCoordinates(Coordinates node1, Coordinates node2, int class_size){
+    public void changeCoordinates(Coordinates node1, Coordinates node2, int class_size) {
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 15]
```diff
@@ -101,7 +136,7 @@ public void changeCoordinates(Coordinates node1, Coordinates node2, int class_si
      * Updates the message image and placing for simulation.
      */
     @Override
-    public void render(GraphicsContext gc){
+    public void render(GraphicsContext gc) {
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 16]
```diff
@@ -112,39 +147,57 @@ public void render(GraphicsContext gc){
         y1 += offset; // Sets an offset from the previous message.
 
         // Checks if we are supposed to be animating the message.
-        if(keepAnimating) {
+        if (keepAnimating) {
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 17]
```diff
@@ -112,39 +147,57 @@ public void render(GraphicsContext gc){
...
             //Sets the message text centered above the "dragon".
-            gc.fillText(this.name, x1+animationBounds, y1 + (this.class_size - 2)); // Message description.
+            gc.fillText(this.name, x1 + animationBounds, y1 + (this.class_size - 2)); // Message description.
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 18]
```diff
@@ -112,39 +147,57 @@ public void render(GraphicsContext gc){
...
 
             // Checks if up image is supposed to be shown and if we are sending a message and not a return.
             if (switchImage && fromNode < toNode) {
                 //sets the dimensions of the dragon according to the current class size.
                 gc.drawImage(dragonMessage, x1 + animationBounds,
-                        y1 + (this.class_size), class_size/messageScale, class_size/messageScale); //State Wings Up.
+                        y1 + (this.class_size), class_size / messageScale, class_size / messageScale); //State Wings Up.
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 19]
```diff
@@ -112,39 +147,57 @@ public void render(GraphicsContext gc){
...
                 switchImage = false;
             }
 
             // Checks if down image is supposed to be shown and if we are sending a message and not a return.
             else if (!switchImage && fromNode < toNode) {
                 //sets the dimensions of the dragon according to the current class size.
                 gc.drawImage(dragonMessage2, x1 + animationBounds,
-                        y1 + (this.class_size), class_size/messageScale, class_size/messageScale); //State Wings Down.
+                        y1 + (this.class_size), class_size / messageScale, class_size / messageScale); //State Wings Down.
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 20]
```diff
@@ -112,39 +147,57 @@ public void render(GraphicsContext gc){
...
+            else if (switchDirection && !switchImage) {
+                //sets the dimensions of the dragon according to the current class size.
+                gc.drawImage(dragonMessage2, x1 + animationBounds,
+                        y1 + (this.class_size), class_size / messageScale, class_size / messageScale); //State Wings Down.
                 switchImage = true;
             }
 
             //Checks if up image is supposed to be shown. if this one is used it is a return message.
-            else if(switchImage){
+            else if (switchImage) {
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 21]
```diff
@@ -112,39 +147,57 @@ public void render(GraphicsContext gc){
...
                 //sets the dimensions of the dragon according to the current class size.
                 gc.drawImage(dragonMessageRev, x1 + animationBounds,
-                        y1 + (this.class_size), class_size/messageScale, class_size/messageScale); //State Wings up.
+                        y1 + (this.class_size), class_size / messageScale, class_size / messageScale); //State Wings up.
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 22]
```diff
@@ -112,39 +147,57 @@ public void render(GraphicsContext gc){
...
                 switchImage = false;
             }
 
             //Checks if down image is supposed to be shown. if this one is used we are sending a return message.
-            else{
+            else {
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 23]
```diff
@@ -112,39 +147,57 @@ public void render(GraphicsContext gc){
...
                 //sets the dimensions of the dragon according to the current class size.
                 gc.drawImage(dragonMessageRev2, x1 + animationBounds,
-                        y1 + (this.class_size), class_size/messageScale, class_size/messageScale); //State Wings Down.
+                        y1 + (this.class_size), class_size / messageScale, class_size / messageScale); //State Wings Down.
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 24]
```diff
@@ -153,18 +206,20 @@ else if(switchImage){
 
     /**
      * Getter method for retrieving the beginning node of a message
+     *
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 25]
```diff
@@ -153,18 +206,20 @@ else if(switchImage){
...
      * @return fromNode
      */
-    public int getFromNode(){
+    public int getFromNode() {
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 26]
```diff
@@ -153,18 +206,20 @@ else if(switchImage){
...
         return fromNode;
     }
 
     /**
      * Getter method for retrieving the end node of a message
+     *
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 27]
```diff
@@ -153,18 +206,20 @@ else if(switchImage){
...
      * @return toNode
      */
-    public int getToNode(){
+    public int getToNode() {
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

File: Application/src/visuals/Message.java [hunk 28]
```diff
@@ -153,18 +206,20 @@ else if(switchImage){
...
         return toNode;
     }
 
-}
+}
```
Comment (timjon) [inline Application/src/visuals/Message.java]: Why is this changed?

**Review by timjon — CHANGES_REQUESTED** [review]

File: Application/src/visuals/Message.java (hunk 7 above)
Comment (ghost) [reply Application/src/visuals/Message.java]: All of those indentation changes happend because of i used the fomatting command in Intelj
Would you like them to be changed back?

File: Application/src/visuals/Message.java (hunk 6 above)
Comment (ghost) [reply Application/src/visuals/Message.java]: Theres is a new self calling case that i added 
It needs the trail as well for full integration
Would you like the commented line to be removed before merging and then immediately after merging? Using a new hotfix branch?

Comment (timjon) [reply Application/src/visuals/Message.java]: Its for a feature thats not implemented yet so you shouldnt have code related to it in your pull request, its going to be fixed when the integration between this feature and the trail feature happens. And that is not now since they havent been integrated at all

File: Application/src/visuals/Message.java (hunk 4 above)
Comment (Rashadse) [reply Application/src/visuals/Message.java]: what would the default case be?

**Review by Sebastian-Fransson — CHANGES_REQUESTED** [review]
Offset might be a problem. Said it in person but blocking here untill done.

**Review by timjon — CHANGES_REQUESTED** [review]
1. The trail does not work with the self call
2. Please use the same offset for everything...

File: Application/src/visuals/Message.java [hunk 29]
```diff
@@ -95,6 +94,39 @@ else if(keepAnimating && node1.getX() > node2.getX()){ // Sending a return messa
...
+                    if ((animationBounds -= (class_size / 6)) < 1) {
+                        keepAnimating = false;
+                    }
+                    break;
+
+                //default case
+                default :
+                    selfCallCounter =1;
+            }
+        }
```
Comment (kosara-g) [inline Application/src/visuals/Message.java:130]: there's got to be a better way to change selfCallCounter; I mean, it works but typically, switch statements aren't used like this; is there any particular reason why you chose to implement it like that?

Comment (ghost) [reply Application/src/visuals/Message.java:130]: @gusgolko i used a switch case to separate the different cases i have which are dependent on each other's sequence. your concern is noted and added to the defect backlog for further improvement.

**Review by timjon — CHANGES_REQUESTED** [review]
Offset is not consistent

**Review by timjon — APPROVED** [review]

**Review by Sebastian-Fransson — APPROVED** [review]
